### ISSUE TITLE
Mac: Folder renames must hydrate children and convert them to full

### ISSUE THREAD

**Issue by wilbaker** [Microsoft/VFSForGit#515]
Currently folder renames expand all children, however, they are not hydrated nor converted to full.  While this is OK if the rename stays within the root, it will break if the rename is to a location outside of the root.

**Comment by wilbaker** [comment]
Update: With the changes in #526 we'd have the option of blocking partial directory renames (which is how ProjFS behaves on Windows)  as we'll know if the PreDelete is for a rename or not.

**Comment by wilbaker** [comment]
Per an offline discussion with @jrbriggs and @jeschu1 we'll aim for the same behavior that we have on Windows, and that is blocking renames of partial folders.

We think it *should* be possible to avoid any recursive walks when making the decision if a folder rename should be allowed assuming that the xattr data allows us to tell if the folder is one that was projected by VFS4G (i.e. it's either an expanded or unexpanded folder).  If a folder was created by a user the rename should be allowed.

**Comment by pmj** [comment]
As far as I can tell, we currently don't set any xattr on directories. So to clarify, are you saying we should now do so to record whether a directory and all its children are partial versus full? In other words, we would probably need a number of expected children, as well as a counter of "full" children, and as long as the 2 do not match, the directory is partial and therefore cannot be renamed? (A subdirectory would increment its parent's counter only when its own counts match.)

I'm not sure this effort is worth it given the likely user confusion it will cause. Is there a way we can enable `git mv` but simply disable all other rename operations?

**Comment by wilbaker** [comment]
> As far as I can tell, we currently don't set any xattr on directories. 

That is unfortunate.  It appears that to support this behavior we'd need an upgrade steps where VFS4G would scan its placeholder directory and add xattrs to the directories.

> So to clarify, are you saying we should now do so to record whether a directory and all its children are partial versus full? In other words, we would probably need a number of expected children, as well as a counter of "full" children, and as long as the 2 do not match, the directory is partial and therefore cannot be renamed? (A subdirectory would increment its parent's counter only when its own counts match.)

The idea was a bit simpler than that.  I had (incorrectly) assumed that directories had xattrs on them and the idea was that if the directory had the placeholder xattr on it then the rename of the directory would be blocked.  This block would happen regardless of whether the folder had been expanded.

> I'm not sure this effort is worth it given the likely user confusion it will cause.  Is there a way we can enable `git mv` but simply disable all other directory rename operations?

I *think* the only things that would be required to support this are:

1. Add an upgrade step that writes xattr data onto all expanded and unexpanded directory placeholders.
2. Update the code that writes directories to start adding the xattr.
3. Block renames of directories that have the xattr data on them.

The concern with supporting directory renames is that it could potentially be a very expensive operation (which is why it's blocked on Windows).  Given that we've been OK so far with this approach on Windows I'm leaning toward keeping the same approach on Mac, but I'd like to hear @jrbriggs and @jeschu1's thoughts as well.

**Comment by pmj** [comment]
Hmm, the thing I don't understand is that when we enumerate directories, we fill them with empty files. (And empty directories unless we're enumerating recursively.) But renaming the directory would only do what the user expected once all of its contents was *hydrated*, except when renaming within a virtualisation root, where the contents could be hydrated at the new location too because the provider will claim responsibility.

So what I don't understand is how the presence vs absence of a simple xattr on each directory would help  with this, because we could only remove that xattr once *all* children have been hydrated. Then you'd need to know whether a newly hydrated file is the last item in its parent directory that was empty, so we can remove the xattr.

**Comment by wilbaker** [comment]
> So what I don't understand is how the presence vs absence of a simple xattr on each directory would help with this, because we could only remove that xattr once all children have been hydrated.

I was thinking we'd match the behavior on Windows, which is that any folder that is/was projected by VFS4G cannot be renamed.  In other words, we'd never remove the xattrs from the folders that VFS4G itself writes.

Users would only be allowed to rename folders that they themselves created.  VFS4G will never write new placeholders inside of a user-created folder, and so it's always safe to allow them to be moved/renamed.  If a user moves a placeholder file into a newly created folder, VFS4G will hydrate it and convert it to full (meaning that it would still be safe to rename the user created folder).

**Comment by pmj** [comment]
> […] any folder that is/was projected by VFS4G cannot be renamed. 
> […]
> Users would only be allowed to rename folders that they themselves created.

Thanks, now I understand what you meant. That certainly sounds reasonable and possible, although as you said, we will need an upgrade step for retrofitting the xattrs.

**Comment by pmj** [comment]
Given how awkward it was already to implement #1208, I'd really like to try and refactor/simplify the kauth handler policy code (#900) before I undertake this change.

**Comment by wilbaker** [comment]
@pmj I was also thinking this issue could be kept on hold until we set a minimum version of Mojave (where we'll have the WILL_RENAME event).  Hopefully that will simplify things as well.

I don't have a feel for how big the refactoring for #900 will be.  Would you guess this would be on the orders of days or work or weeks of work?

**Comment by pmj** [comment]
> I don't have a feel for how big the refactoring for #900 will be. Would you guess this would be on the orders of days or work or weeks of work?

I'd prefer to do this in lots of small steps if at all possible, because there's always active work in this area of code and merge conflicts are going to be a huge pain.

One of my main goals with the refactoring would be to avoid the repeated calls to `TryGetVirtualizationRoot()` (that function may not survive at all in a recognisable form, but basically I want to do the work in that function only once per kauth callback).

My second goal would be to prepare the message(s) to be sent to the provider (if any) in the main part of the function and then actually send it/them at the end. I think it should also be possible to decide the full *sequence* of actions the provider should perform if more than one is required, and it would simplify the code massively to be able to send them to the provider in one message. I think this should be possible to do without causing any TOCTOU style situations as the provider serialises message handling on each file anyway.

I think the first step will be to get a good overview of what pieces of information is required for which decisions, and what the cost of acquiring each piece is. I think from that it'll become clearer what the code *should* look like and then the individual transformations will hopefully also become obvious - and I'll be able to give a sensible estimate rather than a hand-wave.

> @pmj I was also thinking this issue could be kept on hold until we set a minimum version of Mojave (where we'll have the WILL_RENAME event). Hopefully that will simplify things as well.

I don't think the version check adds that much complexity, it's more that the existing code is pretty unreadable and not straightforward to change. We've already got version checks in there, so it'll just become one more item in the giant decision tree.

**Comment by pmj** [comment]
I'm just getting started on implementing this, and step 1 is tagging projected directories by setting an xattr when creating the placeholder. So I need to put together a directory equivalent to `struct PrjFSFileXAttrData`, which has the following layout:

```
struct PrjFSFileXAttrData
{
    PrjFSXattrHeader header;
    
    unsigned char providerId[PrjFS_PlaceholderIdLength];
    unsigned char contentId[PrjFS_PlaceholderIdLength];
};
```

My understanding of git internals is pretty poor, at least compared to everyone else working on VFS4G I'd venture, so this is probably a silly question, but: I'm not quite clear on whether a directory would have a `contentId`/SHA associated with it?

The `providerId` appears to be simply a version number (1) expanded into a 128-byte (!) array. I assume the directory xattr will need this, but I'm not clear on whether I'd need to increment that version number?
